zenoh/src/net/runtime/orchestrator.rs

```diff
@@ -338,72 +341,90 @@ impl Runtime {
+                let retry_config = self.get_connect_retry_config(&endpoint);
+                if retry_config.timeout().is_zero() || self.get_global_connect_timeout().is_zero() {
+                    tracing::debug!(
+                        "Try to connect: {:?}: global timeout: {:?}, retry: {:?}",
+                        endpoint,
+                        self.get_global_connect_timeout(),
+                        retry_config
+                    );
+                    // try to connect directly when there is no timeout configuration
+                    if self.peer_connector(endpoint).await.is_ok() {
+                        success_flag = true;
+                    }
+                } else {
+                    peers_to_retry.push(endpoint);
                 }
-            } else {
-                peers_to_retry.push(endpoint);
             }
-        }
-        // sequentially try to connect to one of the remaining peers
-        // respecting connection retry delays
-        match self.peers_connector_retry(peers_to_retry, true).await {
-            Ok(_) => Ok(()),
-            Err(_) => {
-                let e = zerror!("Unable to connect to any of {:?}! ", peers);
-                tracing::warn!("{}", &e);
-                Err(e.into())
+            // sequentially try to connect to one of the remaining peers
+            // respecting connection retry delays
+            if self
+                .peers_connector_retry(peers_to_retry, false)
+                .await
+                .is_ok()
+            {
+                success_flag = true;
+            }
+            // any endpoint in the group is available, it's marked as success and break
+            if success_flag {
+                break;
             }
         }
+
+        // return error if none of them succeeded
+        if success_flag {
+            Ok(())
+        } else {
+            let e = zerror!("Unable to connect to any of {:?}! ", peers);
+            tracing::warn!("{}", &e);
+            Err(e.into())
+        }
     }
 
-    async fn connect_peers_multiply_links(&self, peers: &[EndPoint]) -> ZResult<()> {
-        for peer in peers {
-            let endpoint = peer.clone();
-            let retry_config = self.get_connect_retry_config(&endpoint);
-            tracing::debug!(
-                "Try to connect: {:?}: global timeout: {:?}, retry: {:?}",
-                endpoint,
-                self.get_global_connect_timeout(),
-                retry_config
-            );
-            if retry_config.timeout().is_zero() || self.get_global_connect_timeout().is_zero() {
-                // try to connect and exit immediately without retry
-                if let Err(e) = self.peer_connector(endpoint).await {
-                    if retry_config.exit_on_failure {
+    async fn connect_peers_multiply_links(&self, peers: &[EndPoints]) -> ZResult<()> {
+        for peer_group in peers {
+            for peer in peer_group.as_vec() {
```

Comment:
That is not the behavior we want. We would like Zenoh to connect to one and only one peer of each group. In case of background connection we want Zenoh to spawn a single task that tries to connect to the group.